### PR TITLE
Rectify image if any distortion coefficients are non-zero

### PR DESCRIPTION
**Pull request by lucasw**
Unit tests included.

For #117

### CODE REVIEW

File: image_geometry/test/utest.cpp [hunk 1]
```diff
@@ -161,6 +162,115 @@ TEST_F(PinholeTest, initialization)
...
+
+  // Test that rectified image the same using empty distortion
+  cam_info_2.D.resize(0);
+  model_.fromCameraInfo(cam_info_2);
+  model_.rectifyImage(distorted_image, rectified_image);
+  cv::absdiff(distorted_image, rectified_image, diff_image);
+  error = cv::sum(diff_image)[0];
+  EXPECT_EQ(error, 0);
+
+  // debug
```
Comment (vrabaud) [inline image_geometry/test/utest.cpp]: This adds an extra dependency on highgui so I guess you can remove it right ?

Comment (lucasw) [inline image_geometry/test/utest.cpp]: I can remove it and the `#include <opencv2/highgui/highgui.hpp>` above.

Comment (lucasw) [inline image_geometry/test/utest.cpp]: It's gone now.

File: image_geometry/test/utest.cpp [hunk 2]
```diff
@@ -161,6 +161,105 @@ TEST_F(PinholeTest, initialization)
...
+  // since a completely different image would be on the order of
+  // width * height * 255 = 78e6
+  const double diff_threshold = 10000.0;
+  double error;
+
+  // Test that rectified image is sufficiently different
+  // using default distortion
+  model_.rectifyImage(distorted_image, rectified_image);
+  cv::absdiff(distorted_image, rectified_image, diff_image);
+  error = cv::sum(diff_image)[0];
```
Comment (vrabaud) [inline image_geometry/test/utest.cpp]: why not use `cv::norm(distorted_image, rectified_image,cv::NORM_L1)` ?

File: image_geometry/test/utest.cpp [hunk 3]
```diff
@@ -161,6 +161,105 @@ TEST_F(PinholeTest, initialization)
...
+  model_.fromCameraInfo(cam_info_2);
+  model_.rectifyImage(distorted_image, rectified_image);
+  cv::absdiff(distorted_image, rectified_image, diff_image);
+  error = cv::sum(diff_image)[0];
+  EXPECT_GT(error, diff_threshold);
+
+  // Test that rectified image the same using zero distortion
+  for (size_t i = 0; i < cam_info_2.D.size(); ++i)
+  {
+    cam_info_2.D[i] = 0.0;
```
Comment (vrabaud) [inline image_geometry/test/utest.cpp]: more C++ to use `cam_info_2.D.assign(cam_info_2.D.size(), 0);`

File: image_geometry/test/utest.cpp [hunk 4]
```diff
@@ -161,6 +161,105 @@ TEST_F(PinholeTest, initialization)
...
+  }
+  model_.fromCameraInfo(cam_info_2);
+  model_.rectifyImage(distorted_image, rectified_image);
+  cv::absdiff(distorted_image, rectified_image, diff_image);
+  error = cv::sum(diff_image)[0];
+  EXPECT_EQ(error, 0);
+
+  // Test that rectified image the same using empty distortion
+  cam_info_2.D.resize(0);
```
Comment (vrabaud) [inline image_geometry/test/utest.cpp]: more c++ to use `cam_info_2.D.clear()`

**Comment by vrabaud** [comment]
looks good overall, please take the comments into account, that just makes less code to review :) And please squash your commits too once done. Thx.

**Comment by lucasw** [comment]
Made those changes, and squashed.

File: image_geometry/test/utest.cpp [hunk 5]
```diff
@@ -161,6 +161,98 @@ TEST_F(PinholeTest, initialization)
...
+  for (size_t x = 0; x <= cam_info_.width; x += cam_info_.width/10)
+  {
+    // draw the lines thick so the prorportion of interpolation error is reduced
+    cv::line(distorted_image,
+             cv::Point(x, 0), cv::Point(x, cam_info_.height),
+             color, type, thickness);
+  }
+
+  cv::Mat rectified_image;
+  cv::Mat diff_image;
```
Comment (vrabaud) [inline image_geometry/test/utest.cpp]: this is useless now right ?

**Comment by lucasw** [comment]
Deleted unused diff image mat.

**Comment by vrabaud** [comment]
Thx !

**Comment by vrabaud** [comment]
Thix closes #117

**Pull request merged**